File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/argument.rb
# frozen_string_literal: true

require 'cucumber/cucumber_expressions/group'
require 'cucumber/cucumber_expressions/errors'

module Cucumber
  module CucumberExpressions
    # An argument extracted from a match
    class Argument
      attr_reader :group, :parameter_type

      # rubocop:disable Metrics/AbcSize
      def self.build(tree_regexp, text, parameter_types)
        group = tree_regexp.match(text)
        return nil if group.nil?

        if group.children.length != parameter_types.length
          # rubocop:disable Layout/LineLength
          raise CucumberExpressionError, "Expression #{tree_regexp.regexp.inspect} has #{group.children.length} capture groups (#{group.children.map(&:value)}), but there were #{parameter_types.length} parameter types (#{parameter_types.map(&:name)})"
          # rubocop:enable Layout/LineLength
        end

        parameter_types.zip(group.children).map do |parameter_type, arg_group|
          Argument.new(arg_group, parameter_type)
        end
      end
      # rubocop:enable Metrics/AbcSize

      def initialize(group, parameter_type)
        @group = group
        @parameter_type = parameter_type
      end

      def value(self_obj = :nil)
        raise 'No self_obj' if self_obj == :nil

        group_values = @group ? @group.values : nil
        @parameter_type.transform(self_obj, group_values)
      end
    end
  end
end

File: cucumber-expressions/go/argument.go
package cucumberexpressions

import "fmt"

type Argument struct {
	group         *Group
	parameterType *ParameterType
}

func BuildArguments(treeRegexp *TreeRegexp, text string, parameterTypes []*ParameterType) []*Argument {
	group := treeRegexp.Match(text)
	if group == nil {
		return nil
	}
	argGroups := group.Children()
	if len(argGroups) != len(parameterTypes) {
		panic(fmt.Errorf("%s has %d capture groups (%v), but there were %d parameter types (%v)", treeRegexp.Regexp().String(), len(argGroups), argGroups, len(parameterTypes), parameterTypes))
	}
	arguments := make([]*Argument, len(parameterTypes))
	for i, parameterType := range parameterTypes {
		arguments[i] = NewArgument(argGroups[i], parameterType)
	}
	return arguments
}

func NewArgument(group *Group, parameterType *ParameterType) *Argument {
	return &Argument{
		group:         group,
		parameterType: parameterType,
	}
}

func (a *Argument) Group() *Group {
	return a.group
}

func (a *Argument) GetValue() interface{} {
	values := a.group.Values()
	if values == nil {
		return nil
	}
	return a.parameterType.Transform(values)
}

func (a *Argument) ParameterType() *ParameterType {
	return a.parameterType
}
